core/consensus/common/src/import_queue.rs

```diff
@@ -97,227 +96,445 @@ pub trait ImportQueue<B: BlockT>: Send + Sync {
+	fn new(
+		result_port: Receiver<BlockImportWorkerMsg<B>>,
+		worker_sender: Sender<BlockImportWorkerMsg<B>>,
+		justification_import: Option<SharedJustificationImport<B>>,
+	) -> Sender<BlockImportMsg<B>> {
+		let (sender, port) = channel::unbounded();
+		let _ = thread::Builder::new()
+			.name("ImportQueue".into())
+			.spawn(move || {
+				let mut importer = BlockImporter {
+					port,
+					result_port,
+					worker_sender,
+					queue_blocks: HashSet::new(),
+					best_importing_number: Zero::zero(),
+					link: None,
+					justification_import,
+				};
+				while importer.run() {
+					// Importing until all senders have been dropped...
+				}
+			})
+			.expect("ImportQueue thread spawning failed");
+		sender
+	}
+
+	fn run(&mut self) -> bool {
+		let msg = select! {
+			recv(self.port) -> msg => {
+				match msg {
+					// Our sender has been dropped, quitting.
+					Err(_) => return false,
+					Ok(msg) => ImportMsgType::FromNetwork(msg)
+				}
+			},
+			recv(self.result_port) -> msg => {
+				match msg {
+					Err(_) => unreachable!("1. We hold a sender to the Worker, 2. it should not quit until that sender is dropped"),
+					Ok(msg) => ImportMsgType::FromWorker(msg),
+				}
+			}
+		};
+		match msg {
+			ImportMsgType::FromNetwork(msg) => self.handle_network_msg(msg),
+			ImportMsgType::FromWorker(msg) => self.handle_worker_msg(msg),
+		}
+	}
+
+	fn handle_network_msg(&mut self, msg: BlockImportMsg<B>) -> bool {
+		match msg {
+			BlockImportMsg::ImportBlocks(origin, incoming_blocks) => {
+				self.handle_import_blocks(origin, incoming_blocks)
+			}
+			BlockImportMsg::Clear => self.handle_clear(),
+			BlockImportMsg::Status(reply_sender) => self.handle_status(reply_sender),
+			BlockImportMsg::IsImporting(hash, reply_sender) => {
+				self.handle_is_importing(hash, reply_sender)
 			}
-			if queue_lock.is_empty() {
-				qdata.signal.wait(&mut queue_lock);
+			BlockImportMsg::ImportJustification(who, hash, number, justification) => {
+				self.handle_import_justification(who, hash, number, justification)
+			}
+			BlockImportMsg::Start(link) => {
+				if let Some(justification_import) = self.justification_import.as_ref() {
+					justification_import.on_start(&*link);
+				}
+				self.link = Some(link);
+			}
+			BlockImportMsg::Stop => {
+				self.handle_clear();
+				return false;
+			}
+		}
+		true
+	}
+
+	fn handle_worker_msg(&mut self, msg: BlockImportWorkerMsg<B>) -> bool {
+		let results = match msg {
+			BlockImportWorkerMsg::Imported(results) => (results),
+			_ => unreachable!("Import Worker does not send ImportBlocks message"),
+		};
+		let mut has_error = false;
+		for (result, hash) in results {
+			self.queue_blocks.remove(&hash);
+
+			if has_error {
+				continue;
 			}
 
-			match queue_lock.pop_front() {
-				Some(new_blocks) => new_blocks,
-				None => break,
+			if result.is_err() {
+				self.best_importing_number = Zero::zero();
+				has_error = true;
 			}
+
+			let link = match self.link.as_ref() {
```

Comment:
Ok so the thing is that a lot of the tests are relying on this, and as this PR plugs the actual `BasicQueue` into them. I've added a `trace` in case this were to actually happen in production...